In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
phone=pd.read_csv(r'C:\Users\hp\AppData\Local\Programs\Python\Python312\Scripts\Anveshanproject\Phone_cleaned_dataset.csv')

In [3]:
phone.head()

Unnamed: 0 Brand me  Ratings  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  \
0           0       LG      4.3  4.0         6.00           48         13   
1           1    IKall      3.4  6.0         4.50           48         12   
2           2    Nokia      4.3  4.0         4.50           64         16   
3           3  Samsung      4.4  6.0         6.40           48         15   
4           4     POCO      4.5  6.0         6.18           35         15   

   Battery_Power    Price  
0           4000  24999.0  
1           4000  15999.0  
2           4000  15000.0  
3           3800  18999.0  
4           3800  18999.0

In [4]:
phone.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 521 entries, 0 to 520
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     521 non-null    int64  
 1   Brand me       521 non-null    object 
 2   Ratings        521 non-null    float64
 3   RAM            521 non-null    float64
 4   Mobile_Size    521 non-null    float64
 5   Primary_Cam    521 non-null    int64  
 6   Selfi_Cam      521 non-null    int64  
 7   Battery_Power  521 non-null    int64  
 8   Price          521 non-null    float64
dtypes: float64(4), int64(4), object(1)
memory usage: 36.8+ KB


In [5]:
phone.shape

(521, 9)

dropping the ratings and unnamed columns

In [6]:
phone=phone.drop('Ratings',axis=1)

In [7]:
phone

Unnamed: 0 Brand me  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  \
0             0       LG  4.0         6.00           48         13   
1             1    IKall  6.0         4.50           48         12   
2             2    Nokia  4.0         4.50           64         16   
3             3  Samsung  6.0         6.40           48         15   
4             4     POCO  6.0         6.18           35         15   
..          ...      ...  ...          ...          ...        ...   
516         516  Samsung  4.0         6.20           35          8   
517         517    Black  6.0         6.39           48         12   
518         518  Karbonn  6.0         4.54           48         12   
519         519  Alcatel  3.0         6.20           48          1   
520         520  Blacear  6.0         4.50           64         15   

     Battery_Power    Price  
0             4000  24999.0  
1             4000  15999.0  
2             4000  15000.0  
3             3800  18999.0  
4             3800  18999.0  
..             ...      ...  
516           3500  53990.0  
517           3800  31999.0  
518           2800   1299.0  
519           3800   9790.0  
520           1050    799.0  

[521 rows x 8 columns]

In [8]:
phone=phone.drop('Unnamed: 0',axis=1)

In [9]:
phone

Brand me  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  Battery_Power    Price
0         LG  4.0         6.00           48         13           4000  24999.0
1      IKall  6.0         4.50           48         12           4000  15999.0
2      Nokia  4.0         4.50           64         16           4000  15000.0
3    Samsung  6.0         6.40           48         15           3800  18999.0
4       POCO  6.0         6.18           35         15           3800  18999.0
..       ...  ...          ...          ...        ...            ...      ...
516  Samsung  4.0         6.20           35          8           3500  53990.0
517    Black  6.0         6.39           48         12           3800  31999.0
518  Karbonn  6.0         4.54           48         12           2800   1299.0
519  Alcatel  3.0         6.20           48          1           3800   9790.0
520  Blacear  6.0         4.50           64         15           1050    799.0

[521 rows x 7 columns]

In [10]:
phone.shape

(521, 7)

In [14]:
X=phone.drop(columns='Price')
y=phone['Price']

In [15]:
X

Brand me  RAM  Mobile_Size  Primary_Cam  Selfi_Cam  Battery_Power
0         LG  4.0         6.00           48         13           4000
1      IKall  6.0         4.50           48         12           4000
2      Nokia  4.0         4.50           64         16           4000
3    Samsung  6.0         6.40           48         15           3800
4       POCO  6.0         6.18           35         15           3800
..       ...  ...          ...          ...        ...            ...
516  Samsung  4.0         6.20           35          8           3500
517    Black  6.0         6.39           48         12           3800
518  Karbonn  6.0         4.54           48         12           2800
519  Alcatel  3.0         6.20           48          1           3800
520  Blacear  6.0         4.50           64         15           1050

[521 rows x 6 columns]

In [16]:
y

0      24999.0
1      15999.0
2      15000.0
3      18999.0
4      18999.0
        ...   
516    53990.0
517    31999.0
518     1299.0
519     9790.0
520      799.0
Name: Price, Length: 521, dtype: float64

In [17]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [19]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline

In [20]:
ohe=OneHotEncoder()

In [21]:
ohe.fit(X[['Brand me']])

OneHotEncoder()

In [23]:
column_trans=make_column_transformer((OneHotEncoder(categories=ohe.categories_),['Brand me']),remainder='passthrough')
lr=LinearRegression()
pipe=make_pipeline(column_trans,lr)
pipe.fit(X_train,y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(categories=[array(['Alcatel', 'Apple', 'Blacear', 'Blacerry', 'Black', 'BlackZone',
       'Doogee', 'Easyfone', 'Forme', 'Good', 'Google', 'Grabo',
       'GreenBerry', 'Hicell', 'Honor', 'Huawei', 'I', 'IKall', 'InFocus',
       'Infinix', 'Inovu', 'Itel', 'Jivi', 'Jmax', 'Karbonn', 'Kechaoda',
       'LG', 'Lava', 'Lenovo', 'MTR', 'Mafe', 'Megus', 'Meizu', 'Mi',
       'Micax', 'Moto', 'Motorola', 'Muphone', 'Nexus', 'Nokia', 'OPPO',
       'OnePlus', 'POCO', 'Peace', 'Q-Tel', 'Realme', 'Redmi', 'Salora',
       'Samsung', 'Ssky', 'Tecno', 'Tork', 'Vivo', 'iQOO'], dtype=object)]),
                                                  ['Brand me'])])),
                ('linearregression', LinearRegression())])

In [24]:
y_pred=pipe.predict(X_test)

In [25]:
r2_score(y_test,y_pred)

0.22797981903094866

In [28]:
scores=[]
for i in range (100):
    X_train, X_test, y_train, y_test=train_test_split(X,y, test_size=0.2, random_state=i)
    column_trans=make_column_transformer((OneHotEncoder(categories=ohe.categories_),['Brand me']),remainder='passthrough')
    lr=LinearRegression()
    pipe=make_pipeline(column_trans,lr)
    pipe.fit(X_train,y_train)
    y_pred=pipe.predict(X_test)
    #print(r2_score(y_test,y_pred),i)
    scores.append(r2_score(y_test,y_pred))

In [29]:
np.argmax(scores)

47

In [30]:
scores[np.argmax(scores)]

0.6993726421235386

This Model has a r2_score of 0.699 at the random_state=47